### ISSUE TITLE
compact style

### ISSUE THREAD

**Issue by romain-noel** [sjelatex/derivative#1]
Hello Simon,
as mentioned on [tex.stackexchange](https://tex.stackexchange.com/questions/652223/write-a-derivative-operator-without-denominator-using-diffcoef) I think it would be very handy to add an option to change the style of derivatives into something compact like `$\pdiv[sytle=compact]{f}{x} = \partial_x f$`. 

However, you wrote that you don't think that is a good idea. 
Can you explain me why ?

**Comment by sjelatex** [comment]
The reason why I don't think it's a good idea is because the current implementation requires using a function that accepts two arguments e.g. `\frac{}{}` - because of the `style-frac` options.

Also I would probably need to make a new set of options just to handle spacing for this case and maybe some other options as well. The package already has a lot of options, and there are cases it has to reset them all (by setting them).  Adding more options would mean it sets a lot of unused options when `style=compact` and when `style=fraction`. It is already a problem because of the star and slash variant of some options. And I don't really see a way out.

And I don't know what the star and slash argument should do if `style=compact`?. I think they would be rendered unusable.

The point of \odif and friends was not only to act as a differential operator in integrals and equations, but also as a shorthand for derivatives. Because of the former I couldn't add a function argument simply because it would have no use. The latter could use one, but I don't really see much of a benefit having it. So I decided to leave the function argument out.

Maybe it is better to make a new concept for this e.g. partial shorthand derivative
    \NewDocumentCommand\psdv{ o m m e{_^} }
    { code implementation }



But I agree. Having the option to switch between fractional and compact versions on the fly would be really nice, without having to change command names, arguments.

**Comment by sjelatex** [comment]
I brainstormed and wrote some thoughts down. It's unstructured rambling, but TLDR I need to think more about it.

The reason why I don't think it's a good idea is because the current implementation requires using a function that accepts two arguments e.g. `\frac{}{}` - because of the `style-frac` options.

Also I would probably need to make a new set of options just to handle spacing for this case and maybe some other options as well. The package already has a lot of options, and there are cases it has to reset them all (by setting them).  Adding more options would mean it sets a lot of unused options when `style=compact` and when `style=fraction`. It is already a problem because of the star and slash variant of some options. And I don't really see a way out.

And I don't know what the star and slash argument should do if `style=compact`?. I think they would be rendered unusable. And which set of options should be used, versions with the star, the slash those without, or new ones?

The point of \odif and friends was not only to act as a differential operator in integrals and equations, but also as a shorthand for derivatives. Because of the former I couldn't add a function argument simply because it would have no use. The latter could use one, but I didn't really see much of a benefit having it. So I decided to leave the function argument out. But I see now it might now have covered all use cases.

Maybe it is better to make a new concept for this e.g. partial shorthand derivative
```
    % #1 = key-value
    % #2 = function
    % #3 = variables
    % #4 = subscript value
    % #5 = superscript value
    \NewDocumentCommand\psdv{ o m m e{_^} }
    { code implementation }
```
It would avoid a lot of the mentioned issues, but would it be distinctive enough to have it's own implementation?

But I agree. Having the option to switch between fractional and compact versions on the fly would be really nice, without having to change command names, arguments etc. But on the other hand, I would prefer having `\pdv` and friends stay as derivative in fractional form. Because, I think it would make the package (hopefully) easier to use for (new) users.

If we said the option was added, I could use an if statement to avoid the two argument function mentioned earlier. But I would probably still need to add some new options though.

It's a dilemma. You really made me think of how to implement a new feature. A tough nut to crack.
I think you said something about helping with the implementation?

**Comment by romain-noel** [comment]
I do understand your arguments and dilemma.

Indeed, by reading a bit more the source code (and documentation) I realized that you require a two arguments command.
Also, I don't see what could be the star and slash variant of a compact style. 
But I agree that, by default, `\odv` and friends should remain in a fractional form.

Just to sum up and try to help your brainstorming.
As you have mentioned, there is 3 (or 4) ways to go with this new compact feature.
1. add an option to `\odv`and friends,
2. add an argument to `\odif` and friends,
3. create new functions `\osdv` and friends,
4. do nothing.

If you agree that the option 4 can be left behind since such feature will be useful, we can look at advantages and disadvantages of the three other options.
Otherwise, I will try to convince you about the usefulness of such feature.
If you see another way to do please let me know.

So advantages and disadvantages:
1. add an option to `\odv`and friends
++ the user can change on the fly the visual aspect of its derivative without changing command names,
++ the number of commands remains limited for the user,
++ the same meaning keeps the same command,
-- we have to add more options and setting to deal with all the cases (more work)
-- the star and slash variants will probably be ignored when the compact style is enabled. 
2. add an argument to `\odif` and friends,
++ it was the initial purpose of these commands (if I understand correctly?) 
-- it implies to change a completely these commands to accept 2 arguments
-- it will require to implement new options 
-- it also implies to adapt when no second argument is given, or if the 2nd argument is mandatory how to write differential in integral operators: it will become messy.
3. create new functions `\osdv` and friends,
++ it does not impact the previous commands
-- it increases the number of commands for users
-- almost all the options and setting will be the same than `\odv`

Therefore, I see a benefice in adding an option to `\odv` and friends, especially because, in my opinion, the major advantage for the user is to focus on the meaning not on the display. 
Then, the latter can be finely tuned or tweaked according to some global or local options. 
This principle is one of the great strengths of LaTeX and in your package I found.

And yes I would be glad if I can help with the implementation, even if I will need a bit of time to completely understand your code.

**Comment by romain-noel** [comment]
Of course, we can discuss more if you want. 
Otherwise, once you have made your mind, do not hesitate to contact me to share/split the workload if you want.

**Comment by sjelatex** [comment]
Ok, so I just finished adding support/features for fractional calculus (still need some testing as I keep finding bugs, see picture... And optimize the ugly code...). And can turn my attention to this.
<img width="165" alt="image" src="https://user-images.githubusercontent.com/68080512/183241885-a169c9e7-d000-4423-8983-9c44078e36ef.png">

The initial goal of `\odif` and friends was both to be used in integrals (`style-notation=multiple`) and as shorthand notation for (`style-notation=single` and `style-notation=mixed`, depending on preferred style). I didn't add a function argument because of the former use case, and believed it was a small loss the the latter use case.

If a compact notation with a function argument is added, I think it is better to add it to `\odv` and friends. And perhaps use `.groups:n` from [l3keys](https://www.ctan.org/pkg/l3kernel) to solve the more options problem. Could also solve the current one. I have looked at several times over the years, but I haven't found a good example of how to use it. 

Else a optional argument could be added to `\odif` and friends as the last argument to handle a function input.

But I think the former solution sounds better.
